src/picosvg/svg.py

```diff
@@ -679,17 +702,89 @@ def remove_attributes(self, names, xpath="/svg:svg", inplace=False):
 
         return self
 
+    def _select_gradients(self):
+        return self.xpath(" | ".join(f"//svg:{tag}" for tag in _GRADIENT_CLASSES))
+
+    def _collect_gradients(self, inplace=False):
+        if not inplace:
+            svg = SVG(copy.deepcopy(self.svg_root))
+            svg._collect_gradients(inplace=False)
+            return svg
+
+        # Collect gradients; remove other defs
+        defs = etree.Element(f"{{{svgns()}}}defs", nsmap=self.svg_root.nsmap)
+        for gradient in self._select_gradients():
+            gradient.getparent().remove(gradient)
+            defs.append(gradient)
+
+        for def_el in [e for e in self.xpath("//svg:defs")]:
+            def_el.getparent().remove(def_el)
+
+        self.svg_root.insert(0, defs)
+
+    def _apply_gradient_translation(self, inplace=False):
+        if not inplace:
+            svg = SVG(copy.deepcopy(self.svg_root))
+            svg._apply_gradient_translation(inplace=True)
+            return svg
+
+        for el in self._select_gradients():
+            gradient = _GRADIENT_CLASSES[strip_ns(el.tag)].from_element(
+                el, self.view_box()
+            )
+            affine = gradient.gradientTransform
+            a, b, c, d, dx, dy = affine
+            if (dx, dy) == (0, 0):
+                continue
+            affine_prime = affine._replace(e=0, f=0)
+
+            for x_attr, y_attr in _GRADIENT_COORDS[strip_ns(el.tag)]:
+                # if at default just ignore
+                if x_attr not in el.attrib and y_attr not in el.attrib:
+                    continue
+                x = getattr(gradient, x_attr)
+                y = getattr(gradient, y_attr)
+
+                # We need x`, y` such that matrix a b c d 0 0 yields same
+                # result as x,y with a b c d e f
+                # That is:
+                # 1)  ax` + cy` + 0 = ax + cy + e
+                # 2)  bx` + dy` + 0 = bx + dy + f
+                #                   ^ rhs is a known scalar; we'll call r1, r2
+                # multiply 1) by b/a so when subtracted from 2) we eliminate x`
+                # 1)  bx` + (b/a)cy` = (b/a) * r1
+                # 2) - 1)  bx` - bx` + dy` - (b/a)cy` = r2 - (b/a) * r1
+                #         y` = (r2 - (b/a) * r1) / (d - (b/a)c)
+                r1, r2 = affine.map_point((x, y))
+                assert r1 == a * x + c * y + dx
+                assert r2 == b * x + d * y + dy
+                y_prime = (r2 - r1 * b / a) / (d - b * c / a)
+
+                # Sub y` into 1)
+                # 1) x` = (r1 - cy`) / a
+                x_prime = (r1 - c * y_prime) / a
+
+                # sanity check: a`(x`, y`) should be a(x, y)
+                p = affine.map_point((x, y))
```

Comment:
nit: `p` is the same as `(r1, r2)` above, no need to recompute here

Reply:
Ty, fixing